packages/app/app/actions/downloads.ts

```diff
@@ -170,3 +170,34 @@ export const clearFinishedDownloads = createStandardAction(DownloadActionTypes.C
       payload: filteredTracks
     };
   });
+
+/**
+* Helper function to read the old track format into the new format.
+*
+* `Track.artist` and `Track.extraArtists` are written into {@link Track.artists}
+*/
+function getDownloadsBackwardsCompatible(): Download[] {
+  const downloads: Download[] = store.get('downloads');
+
+  downloads.forEach(download => {
+    // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
+    if (download.track.artists || !download.track.artist) {
+      return;
+    }
+
+    // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
+    if (download.track.artist) {
+      // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
+      download.track.artists = _.isString(download.track.artist) ? [download.track.artist] : [download.track.artist.name];
+    }
+
+    // Assuming we have `extraArtists` on a track, we must had an `artist` which
+    // was already saved into `artists`, so this `track.artists` shouldn't be undefined
+    // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
+    download.track.extraArtists?.forEach(artist => {
+      download.track.artists.push(artist);
+    });
+  });
+
+  return downloads;
+}
```

Comment:
This function does the exact same thing as the one in `playlists.ts` and almost the same as the one in `favorites.ts`. The common parts of this logic should be extracted and can live in e.g. `packages/app/app/actions/helpers.ts`.

Reply:
Not quite. All three are slightly different:

~~~ ts
Download[]: { track: Track }[]
Playlist[]: { tracks: Track[] }[]
Favorites: { tracks: Track[] }
~~~

I tried combining all into a generic function but wasn't successful with it.

<details>
<summary>Just for laughs</summary>

~~~ ts
export interface TrackMember {
  track: Track;
}

export interface TracksMember {
  tracks: Track[];
}

export interface Favorites {
  albums: Album[];
  artists: Artist[];
  tracks: Track[];
}

function isTracksMember(config: TrackMember | TracksMember): config is TracksMember {
  return (config as TracksMember).tracks !== undefined;
}

function isTrackMember(config: TrackMember | TracksMember): config is TrackMember {
  return (config as TrackMember).track !== undefined;
}

function isTracksArrayMember(config: TrackMember | TracksMember | Favorites): config is Favorites {
  return (config as Favorites).tracks !== undefined && (config as Favorites).tracks instanceof Array;
}

function rewriteTrackArtists(track: Track) {
  // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
  if (track.artists || !track.artist) {
    // New format already present, do nothing
    return;
  }

  // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
  track.artists = _.isString(track.artist) ? [track.artist] : [track.artist.name];

  // @ts-expect-error For backwards compatibility we're trying to parse an invalid field
  track.artists = track.artists.concat(track.extraArtists?.map(artist));
}

/**
+* Helper function to read the old track format into the new format.
+*
+* `Track.artist` and `Track.extraArtists` are written into {@link Track.artists}
+*/
export function getFromStoreBackwardsCompatible<T extends Favorites | TrackMember[] | TracksMember[]>(configName: 'downloads' | 'favorites' | 'playlists'): T {
  let config: T = store.get(configName);

  if (config instanceof Array) {
    // Playlist[] || Download[]
    config = config.map((member: TrackMember | TracksMember) => {
      if (isTrackMember(member)) {
        // Download[]
        rewriteTrackArtists(member.track);
      } else if (isTracksMember(member)) {
        // Playlist[]
        member.tracks.forEach(track => {
          rewriteTrackArtists(track);
        });
      }
      return member;
    });
  } else if (isTracksArrayMember(config)) {
    // Favorites: { tracks: Track[] }
    config.tracks = config.tracks.map(track => {
      rewriteTrackArtists(track);
      return track;
    });
    config.albums = config.albums.map(album => {
      album.tracklist = album.tracklist?.map(track => {
        rewriteTrackArtists(track);
        return track;
      });
      return album;
    });
  }

  return config;
}
~~~

</details>